File: modules/research.py
"""
模块一：选题搜索 (research)
数据源：荆州新闻网 jznews.com.cn
"""

import requests
from datetime import datetime, timedelta
from bs4 import BeautifulSoup
from typing import List, Dict


def research(keyword: str, days: int = 7) -> dict:
    """
    搜索热门选题

    Args:
        keyword: 选题关键词
        days: 搜索最近多少天的内容

    Returns:
        dict with status, topics list
    """
    topics = []

    try:
        # 方案一：抓取荆州新闻网首页头条
        topics = _fetch_jznews_homepage(keyword, days)
    except Exception as e:
        return {
            "status": "error",
            "message": f"抓取失败: {str(e)}",
            "topics": []
        }

    if not topics:
        return {
            "status": "ok",
            "message": "未找到相关选题，请尝试更通用的关键词",
            "topics": []
        }

    return {
        "status": "ok",
        "message": f"找到 {len(topics)} 条相关选题",
        "topics": topics
    }


def _fetch_jznews_homepage(keyword: str, days: int) -> List[Dict]:
    """抓取荆州新闻网首页，筛选含关键词的条目"""
    url = "https://www.jznews.com.cn/"
    headers = {
        "User-Agent": (
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
            "AppleWebKit/537.36 (KHTML, like Gecko) "
            "Chrome/120.0.0.0 Safari/537.36"
        ),
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language": "zh-CN,zh;q=0.9",
    }

    resp = requests.get(url, headers=headers, timeout=10)
    resp.encoding = "utf-8"
    soup = BeautifulSoup(resp.text, "html.parser")

    cutoff = datetime.now() - timedelta(days=days)
    results = []

    # 遍历所有新闻链接
    for a in soup.find_all("a", href=True):
        href = a.get("href", "")
        text = a.get_text(strip=True)

        # 过滤：关键词匹配 + 有效长度
        if keyword and keyword not in text:
            continue
        if len(text) < 10 or len(text) > 80:
            continue
        if not any(ext in href for ext in [".html", "/news/", "/content/"]):
            continue

        # 尝试找摘要
        parent = a.find_parent(["li", "div", "p"])
        summary = ""
        if parent:
            ps = parent.find_all("p")
            for p in ps:
                p_text = p.get_text(strip=True)
                if 20 < len(p_text) < 200 and p_text != text:
                    summary = p_text
                    break

        results.append({
            "title": text,
            "url": href if href.startswith("http") else f"https://www.jznews.com.cn{href}",
            "summary": summary,
            "source": "荆州新闻网",
            "score": _calc_score(text, keyword),
            "tag": _extract_tag(text)
        })

    # 去重 + 排序
    seen = set()
    unique = []
    for r in results:
        if r["title"] not in seen:
            seen.add(r["title"])
            unique.append(r)

    unique.sort(key=lambda x: x["score"], reverse=True)
    return unique[:20]  # 最多返回20条


def _calc_score(title: str, keyword: str) -> int:
    """计算选题相关度评分"""
    score = 50
    title_lower = title.lower()
    keyword_lower = keyword.lower()

    if keyword_lower in title_lower:
        score += 30
    if any(w in title_lower for w in ["重磅", "最新", "突发", "刚刚", "首次"]):
        score += 10
    if any(w in title_lower for w in ["数据", "报告", "政策", "规划", "亿元", "项目"]):
        score += 15
    if any(w in title_lower for w in ["突破", "增长", "创新", "第一"]):
        score += 10

    return min(score, 100)


def _extract_tag(title: str) -> str:
    """从标题提取分类标签"""
    tags = {
        "文旅": ["文旅", "旅游", "景区", "演出", "赛事", "文化", "博物馆"],
        "民生": ["高考", "中考", "教育", "医疗", "社保", "就业", "房价"],
        "政策": ["政策", "规划", "公文", "通知", "会议", "领导"],
        "经济": ["经济", "产业", "招商", "项目", "企业", "投资", "金融"],
        "农业": ["农业", "乡村", "农村", "振兴", "粮食", "生态"],
        "科技": ["科技", "数字", "智能", "创新", "研发", "工业"],
    }
    for tag, keywords in tags.items():
        if any(kw in title for kw in keywords):
            return tag
    return "综合"

File: api/research.js
/**
 * research.js · 纯 Node.js 选题搜索
 * 数据源：搜狗微信（无需登录）+ 关键词搜索
 * Vercel Node.js Serverless Function
 */

// ── 评分函数 ─────────────────────────────────────────────────
function parseBody(raw) {
  if (!raw) return {};
  if (typeof raw === "string") return JSON.parse(raw);
  if (Buffer.isBuffer(raw)) return JSON.parse(raw.toString());
  if (typeof raw === "object") return raw;
  return {};
}

function calcScore(title, keyword) {
  const kws = keyword.toLowerCase().split(/[\s,，、]+/);
  const t = title.toLowerCase();
  let score = 40;
  for (const kw of kws) {
    if (t.includes(kw)) score += 15;
  }
  // 标题长度适中得分更高
  if (title.length >= 10 && title.length <= 30) score += 10;
  else if (title.length > 30) score += 5;
  // 数字加分（数据/日期）
  if (/\d+%|\d+万|\d+亿/.test(title)) score += 8;
  return Math.min(score, 95);
}

// ── 标签提取 ─────────────────────────────────────────────────
function extractTag(title) {
  const map = {
    '农业|农村|农产|农机|乡村振兴': '三农',
    '招商|签约|投产|开工|项目': '产业',
    '旅游|景区|游客|文旅|打卡|演出': '文旅',
    '政策|规划|方案|意见|通知': '政策',
    '经济|工业|GDP|增速': '经济',
    '教育|学校|高考|中考|开学': '教育',
    '交通|高铁|高速|机场|大桥|道路': '交通',
    '医疗|医院|医保|健康': '民生',
    'AI|人工智能|机器人|智造': '科技',
    '天气|气温|降雨|防汛': '气象',
    '活动|开幕|论坛|峰会|发布': '活动',
    '降价|优惠|免费|福利': '消费',
  };
  for (const [pattern, tag] of Object.entries(map)) {
    if (new RegExp(pattern).test(title)) return tag;
  }
  return '综合';
}

// ── 搜索搜狗微信 ─────────────────────────────────────────────
async function searchSogou(keyword, days = 7) {
  const url = `https://weixin.sogou.com/weixin?type=2&query=${encodeURIComponent(keyword)}&ie=utf8`;
  const resp = await fetch(url, {
    timeout: 10000,
    headers: {
      'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
      'Accept': 'text/html,application/xhtml+xml',
    },
  });
  const html = await resp.text();

  const results = [];
  // 解析搜狗微信搜索结果
  const itemRegex = /<div class="txt-box">([\s\S]*?)<\/div>\s*<\/div>/g;
  let match;
  while ((match = itemRegex.exec(html)) !== null && results.length < 10) {
    const block = match[1];
    const titleMatch = block.match(/<h3[^>]*>\s*<a[^>]*>([\s\S]*?)<\/a>/);
    const pMatch = block.match(/<p[^>]*>([\s\S]*?)<\/p>/);
    const hrefMatch = block.match(/href="(https?:\/\/[^"]+)"/);

    if (titleMatch) {
      const rawTitle = titleMatch[1].replace(/<[^>]+>/g, '').trim();
      const summary = pMatch ? pMatch[1].replace(/<[^>]+>/g, '').trim().slice(0, 200) : '';
      const href = hrefMatch ? hrefMatch[1] : '';

      if (rawTitle.length >= 8) {
        results.push({
          title: rawTitle,
          url: href,
          summary,
          source: '搜狗微信',
          tag: extractTag(rawTitle),
          score: calcScore(rawTitle, keyword),
          date: new Date().toISOString().slice(0, 10),
        });
      }
    }
  }

  // 备选：用 Bing 搜索
  if (results.length < 3) {
    try {
      const bingUrl = `https://www.bing.com/search?q=${encodeURIComponent(keyword + ' site:mp.weixin.qq.com')}&first=0&count=10`;
      const bingResp = await fetch(bingUrl, {
        timeout: 10000,
        headers: {
          'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
        },
      });
      const bingHtml = await bingResp.text();
      const bingItems = bingHtml.match(/<li class="b_algo"[\s\S]*?<\/li>/g) || [];
      for (const item of bingItems.slice(0, 5)) {
        const titleM = item.match(/<h2[^>]*>([\s\S]*?)<\/h2>/);
        const hrefM = item.match(/href="(https?:\/\/[^"]+)"/);
        const pM = item.match(/<p[^>]*>([\s\S]*?)<\/p>/);
        if (titleM && hrefM) {
          const rawTitle = titleM[1].replace(/<[^>]+>/g, '').trim();
          const summary = pM ? pM[1].replace(/<[^>]+>/g, '').trim().slice(0, 150) : '';
          if (rawTitle.length >= 8) {
            results.push({
              title: rawTitle,
              url: hrefM[1],
              summary,
              source: 'Bing搜索',
              tag: extractTag(rawTitle),
              score: calcScore(rawTitle, keyword),
              date: new Date().toISOString().slice(0, 10),
            });
          }
        }
      }
    } catch (e) {
      console.warn('Bing搜索失败:', e.message);
    }
  }

  // 按评分排序，去重
  const seen = new Set();
  const unique = results.filter(r => {
    if (seen.has(r.title)) return false;
    seen.add(r.title);
    return true;
  });

  unique.sort((a, b) => b.score - a.score);
  return { status: 'ok', keyword, count: unique.length, topics: unique.slice(0, 10) };
}

// ── 主函数 ───────────────────────────────────────────────────
export default async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  if (req.method === 'OPTIONS') return res.status(200).end();

  let body = parseBody(req.body);

  const keyword = body.keyword || '荆州';
  const days = parseInt(body.days) || 7;

  try {
    const result = await searchSogou(keyword, days);
    return res.status(200).json(result);
  } catch (e) {
    console.error('搜索失败:', e.message);
    return res.status(200).json({
      status: 'ok',
      keyword,
      count: 0,
      topics: [{
        title: `${keyword} 最新资讯`,
        url: '',
        summary: '网络搜索暂时不可用，请检查网络后重试。',
        source: '系统',
        tag: '综合',
        score: 50,
        date: new Date().toISOString().slice(0, 10),
      }],
      error: e.message,
    });
  }
}
